### ISSUE TITLE
TypeScript: Zone#create should have required parameter of type CreateZoneRequest

### ISSUE THREAD

**Issue by carnesen** [googleapis/nodejs-dns#126]
Currently the TypeScript type of `Zone#create` is inherited from [`ServiceObject#create`](https://github.com/googleapis/nodejs-common/blob/e1f4ee84a97a4a64ab392a55d29f21a2486af027/src/service-object.ts#L203). Effectively the first argument is optional and if present can have any shape. (TypeScript doesn't do the excess property check on assignments to an empty object type.) That doesn't jibe with either [the docs](https://github.com/googleapis/nodejs-dns/pull/125) or [the runtime checks](https://github.com/googleapis/nodejs-dns/blob/d07673159609f2240394cf2a75a734f9ff546b08/src/index.ts#L231). Instead, the first parameter of `Zone#create` should be required of type `CreateZoneRequest`. (Maybe without the "name" property?).

**Comment by carnesen** [comment]
I'm working on this now. Feel free to assign if that's useful.

**Comment by carnesen** [comment]
Actually it's not obvious to me how to proceed here. `Zone`'s `create` method is inherited from `ServiceObject`, which knows nothing about `dnsName`s. I tried to add some appropriate `create` overloads to the `Zone` class definition, but encountered `Property 'create' in type 'Zone' is not assignable to the same property in base type 'ServiceObject'.`

**Comment by stephenplusplus** [comment]
The way to call our `child.create()` is with the arguments the `parent.createChild()` method expects, *except* the first argument, which is always the name.

```js
const zone = dns.zone('...')
zone.create(config, callback)
// ===
dns.createZone('...', config, callback)
```

The `create` behavior is defined in [`ServiceObject.prototype.create`](https://github.com/googleapis/nodejs-common/blob/65b5f99c61ef48774826432bd8bc8a87248fe68c/src/service-object.ts#L210), which calls `config.createMethod`:

https://github.com/googleapis/nodejs-dns/blob/d07673159609f2240394cf2a75a734f9ff546b08/src/zone.ts#L345

A condensed version of all of this functionality:

```js
class ServiceObject {
  constructor(cfg) {
    this.id = cfg.id
    this.createMethod = cfg.createMethod
  }

  create() {
    const args = [this.id]
    // figure out any additional arguments to provide to
    // the create method, i.e. args.push(options)
    this.createMethod.apply(null, args)
  }
}

class DNS extends Service {
  createZone(name, config, callback) {
    config.name = name
    this.request(...)
  }
}

class Zone extends ServiceObject {
  constructor(parentInstance) {
    super({
      methods: {
        create: true
      },
      createMethod: parentInstance.createZone.bind(parentInstance)
    })
  }
}
```

**Comment by carnesen** [comment]
@stephenplusplus TIL about embedded code snippits. NEAT! The JavaScripty part of all this is working just fine. It's the types that aren't correct. Here is a simplification that illustrates the issue:

```TypeScript
class Service {}

interface CreateZoneOptions {
  dnsName: string;
}

class DNS extends Service {
  createZone(options: CreateZoneOptions) {
    if (typeof options === 'undefined') {
      // This code should never be reached by a TypeScript user.
      // Said another way, the type of "options" here is "never".
      // But let's be kind to our JavaScript users and fail fast.
      throw new Error('Expected argument "options"');
    }
    const { dnsName } = options;
    if (typeof dnsName === 'undefined') {
      // This code should never be reached by a TypeScript user.
      // Said another way, the type of "dnsName" here is "never".
      // But let's be kind to our JavaScript users and fail fast.
      throw new Error('Expected argument "options" to include "dnsName"');
    }
    // Do request to /managedZones and so forth ...
  }
}

interface ServiceObjectOptions {
  createMethod: Function;
}

class ServiceObject {
  createMethod: Function;
  constructor(opts: ServiceObjectOptions) {
    this.createMethod = opts.createMethod;
  }

  create(opts?: any) {
    this.createMethod.apply(null, opts);
  }
}

class Zone extends ServiceObject {
  constructor(dns: DNS) {
    super({ createMethod: dns.createZone.bind(dns) });
  }
}

const dns = new DNS();
const zone = new Zone(dns);

// For "dns", everything works as expected.

// Compiler error "Expected 1 arguments, but got 0."
dns.createZone(); 

// Compiler error: "'dnsLameTypo' does not exist in type 'CreateZoneOptions'."
dns.createZone({ dnsLameTypo: 'carnesen.com' });

// OK!
dns.createZone({ dnsName: 'carnesen.com' });

// For "zone", the typings are not sufficiently strict.

// The following is NOT currently a compiler error. It should be.
// Instead we encounter a runtime error 'Expected argument "options"'
zone.create();

// The following is NOT currently a compiler error. It should be.
// Instead we encounter a runtime error 'Expected argument "options" to include "dnsName"'
zone.create({ dnsLameTypo: 'carnesen.com' });

// OK!
zone.create({ dnsName: 'carnesen.com'});
```

**Comment by carnesen** [comment]
To continue that example, here is one way we could approach it using TypeScript generics:

```TypeScript
interface CreateMethod<T> {
  (opts: T): void;
}

interface BetterServiceObjectOptions<T> {
  createMethod: CreateMethod<T>;
}

class BetterServiceObject<T> {
  private createMethod: CreateMethod<T>;
  constructor(opts: BetterServiceObjectOptions<T>) {
    this.createMethod = opts.createMethod;
  }
  create(opts: T) {
    this.createMethod(opts);
  }
}

class BetterZone extends BetterServiceObject<CreateZoneOptions> {
  constructor(dns: DNS) {
    super({ createMethod: dns.createZone.bind(dns) });
  }
}

const betterZone = new BetterZone(dns);

// Compiler error "Expected 1 arguments, but got 0."
betterZone.create();

// Compiler error "Property 'dnsName' is missing in type '{}'."
betterZone.create({});

// Compiler error "'dnsLameTypo' does not exist in type 'CreateZoneOptions'"
betterZone.create({ dnsLameTypo: 'carnesen.com' });

// OK!
betterZone.create({ dnsName: 'carnesen.com'});
```

**Comment by bcoe** [comment]
@carnesen I know it's been quite some time since you filed this bug, but it should now be addressed in `@google-cloud/dns@1.2.4`, mind taking this for a spin and letting us know if it addresses your issues?

**Comment by carnesen** [comment]
@bcoe Indeed `zone.create` now expects a `config` object of type `CreateZoneRequest` like `dns.createZone` does! See also though https://github.com/googleapis/nodejs-dns/issues/309
